File: lua/problem_80.lua
local M = {}

local function deepest(node)
    if not node then return nil, 0 end
    local lv, ld = deepest(node.left)
    local rv, rd = deepest(node.right)
    if ld > rd then return lv, ld + 1
    elseif rd > ld then return rv, rd + 1
    else return node.val, 1 end
end

function M.deepest_node(root) return deepest(root) end

function M.problem80()
    print("Problem 80:")
    local root = {val='a',
        left  = {val='b', left={val='d',left=nil,right=nil}, right=nil},
        right = {val='c', left=nil, right=nil},
    }
    local v = M.deepest_node(root)
    assert(v == 'd')
    print("All Tests Passed")
end

return M

File: typescript/problem_80.ts
class TreeNode80ts { constructor(public val:string, public left:TreeNode80ts|null=null, public right:TreeNode80ts|null=null){} }

export function deepestNode(root: TreeNode80ts|null): [string, number] {
    if (!root) return ['', 0];
    const [lv,ld] = deepestNode(root.left), [rv,rd] = deepestNode(root.right);
    if (ld > rd) return [lv, ld+1];
    if (rd > ld) return [rv, rd+1];
    return [root.val, 1];
}

export function problem80(): void {
    console.log("Problem 80:");
    const root = new TreeNode80ts('a', new TreeNode80ts('b', new TreeNode80ts('d')), new TreeNode80ts('c'));
    if (deepestNode(root)[0] !== 'd') throw new Error("assertion failed");
    console.log("All Tests Passed");
}
